supporting-blog-content/evaluating-search-relevance-part-1/Retrieve and rerank.ipynb

```diff
@@ -0,0 +1,801 @@
+    "reranking_model = CrossEncoder(\"cross-encoder/ms-marco-MiniLM-L-6-v2\", max_length=512)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b87d87cd-4c87-4ec7-b3d5-e69aa3433f65",
+   "metadata": {},
+   "source": [
+    "Some helper structures to speed up processing"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "22040c13-4040-4ddc-b1e1-ec05662fd64b",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "queries_dict = {q[\"_id\"]: q[\"text\"] for q in queries}\n",
+    "corpus_dict = {doc[\"_id\"]: f\"{doc['title']} {doc['text']}\" for doc in corpus}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f11f5dbc-aa54-4a78-a61a-fa67b2742bf1",
+   "metadata": {},
+   "source": [
+    "and now it's time for the reranking part"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "15a48d1d-e295-43ff-a9ed-273d3f21a9e4",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "results_after_reranking = dict()\n",
+    "\n",
+    "for qid, bm25_res in tqdm(bm25_responses.items(), total=len(bm25_responses)):\n",
+    "\n",
+    "    query_text = queries_dict[qid]\n",
+    "    doc_ids = [doc_id for doc_id, _ in bm25_res.items()]\n",
+    "    if len(doc_ids) == 0:\n",
+    "        results_after_reranking[qid] = dict()\n",
+    "        continue\n",
+    "\n",
+    "    doc_texts = [corpus_dict[doc_id] for doc_id in doc_ids]\n",
+    "\n",
+    "    # rescore with the reranking model\n",
+    "    scores = reranking_model.predict([(query_text, doc_text) for doc_text in doc_texts])\n",
+    "\n",
+    "    results_after_reranking[qid] = {\n",
+    "        doc_id: float(score) for doc_id, score in zip(doc_ids, scores)\n",
+    "    }"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a85e0f51-fedc-4490-800c-48fd268d8db7",
+   "metadata": {},
+   "source": [
+    "and let's calculate the metric scores for the reranked results"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1c145552-a372-4624-92f8-4798636dd3ec",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "post_reranking_eval_scores_per_query = evaluator.evaluate(results_after_reranking)\n",
+    "\n",
+    "post_reranking_eval_scores = defaultdict(list)\n",
+    "\n",
+    "for qid, vals in post_reranking_eval_scores_per_query.items():\n",
+    "    for metric, metric_score in vals.items():\n",
+    "        post_reranking_eval_scores[metric].append(metric_score)\n",
+    "\n",
+    "for metric, scores in post_reranking_eval_scores.items():\n",
+    "    print(f\"{metric}: {np.mean(scores)}\")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "d35c3a11-9d85-4b5b-9c68-0add866f3700",
+   "metadata": {},
+   "source": [
+    "which in most cases will provide a significant boost in performance"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a5f50faf-9192-43b6-be35-6700b740881d",
+   "metadata": {},
+   "source": [
+    "## Bonus section\n",
```

Comment:
It could be nice to explain why looking at Judge_rate is interesting, and what it tells you about your dataset.